### PR TITLE
Reduce chunk size

### PR DESCRIPTION
**Pull request by cosmo0920**
* tag separated chunk
* handle old flat buffer style conf with compat_parameters plugin helper
* add test case for invalid chunk_keys

### CODE REVIEW

File: lib/fluent/plugin/out_memcached.rb [hunk 1]
```diff
@@ -13,15 +17,22 @@ class Fluent::Plugin::MemcachedOutput < Fluent::Plugin::Output
   config_param :value_format, :string, :default => 'raw'
   config_param :param_names, :string, :default => nil # nil doesn't allowed for json
 
+  config_section :buffer do
+    config_set_default :@type, DEFAULT_BUFFER_TYPE
+    config_set_default :chunk_keys, ['tag']
```
Comment (innossh) [inline lib/fluent/plugin/out_memcached.rb]: not to set default

File: lib/fluent/plugin/out_memcached.rb [hunk 2]
```diff
@@ -13,15 +17,22 @@ class Fluent::Plugin::MemcachedOutput < Fluent::Plugin::Output
...
   attr_accessor :formatter
 
   def configure(conf)
+    compat_parameters_convert(conf, :buffer)
     super
     if @value_format == 'json' and @param_names.nil?
       raise Fluent::ConfigError, "param_names MUST be specified in the case of json format"
     end
     @formatter = RecordValueFormatter.new(@increment, @value_separater, @value_format, @param_names)
+    raise Fluent::ConfigError, "'tag' in chunk_keys is required." unless @chunk_key_tag
```
Comment (innossh) [inline lib/fluent/plugin/out_memcached.rb]: it doesn't need to check `@chunk_key_tag`

**Review by innossh — COMMENTED** [review]
@cosmo0920 Thanks for the PR. It looks nice. And I said, "It's better to reduce the chunk size", but I thought better of it. 🙇🙇🙇
For the plugin users, it may be better to be able to freely choose the chunk keys than to reduce chunk size. We can do that because the `write` function in this plugin doesn't use `tag`, right? How do you think about this?

File: lib/fluent/plugin/out_memcached.rb (hunk 1 above)
Comment (cosmo0920) [reply lib/fluent/plugin/out_memcached.rb]: This is a default value for chunk_keys.
Unless this value is not set, tag separated chunk does not work.

File: lib/fluent/plugin/out_memcached.rb (hunk 2 above)
Comment (cosmo0920) [reply lib/fluent/plugin/out_memcached.rb]: This is the internal instance variable for existence checking for tag in chunk_keys.
See: https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/output.rb#L278

Comment (cosmo0920) [reply lib/fluent/plugin/out_memcached.rb]: Ah, I've got what you said.
I'll fix it.

**Comment by cosmo0920** [comment]
> We can do that because the write function in this plugin doesn't use tag, right? How do you think about this?

Right. I've overlooked that tag is unnecessary in `#write`.
With this PR, users will be a be to choose just buffered plugin[1] or tag separated chunk buffered plugin[2] or time sliced buffed output plugin[3].

[1]
```aconf
<match **>
  @type memcached
   <buffer>
      @type memory
   </buffer>
   # <snip>
</match>
```

[2]
```aconf
<match **>
  @type memcached
   <buffer tag>
      @type memory
   </buffer>
   # <snip>
</match>
```

[3]
```aconf
<match **>
  @type memcached
   <buffer tag, time>
      @type memory
      timekey 3600 # for 1 hour
   </buffer>
   # <snip>
</match>
```

**Pull request merged**